codechain/primitives/PlatformAddress.py

```diff
@@ -0,0 +1,114 @@
+import re
+
+from ..crypto import bech32_decode
+from ..crypto import bech32_encode
+from ..crypto import blake160
+from .HexString import H160
+from .HexString import H512
+
+
+class PlatformAddress:
```

Comment:
PlatformAddress should be a value type. Inherit `typing.NamedTuple`.

Reply:
what is the value type?

Reply:
A value type is a type that can be compared by values of fields, not by the reference.

```python
def foo():
	return 10000000 + 10000000
def bar():
	return 20000000
a = foo()
b = bar()
assert a is not b
# they don't have the same reference 
# a, b, they are different objects that are allocated in different heap locations.
# Python's implementation detail:
# Python's integer is also an object. However, 
# Python caches and shares small integer reference, 
# but not for the large integers. They allocate it every time. 
# see: https://stackoverflow.com/questions/15171695/whats-with-the-integer-cache-maintained-by-the-interpreter)
assert a == b
# but they are compared by its value. 
# If the value is same, `__eq__` returns `true`, and `__hash__` is the same.
```

```python
class ComplexType:
	def __init__(self, value):
		self.value = value
	def __add__(self, other):
		return Wrapper(self.value + other.value)
def foo():
	return ComplexType(20000000)
def bar():
	return ComplexType(20000000)
a = foo()
b = bar()
assert a is not b 
# They are different entities that are allocated in different locations.
assert a != b 
# they are compared by the reference, even its fields are identical values.
```
To make `ComplexType` be compared by value, then we should implement `__eq__` and `__hash__`. Python already did it to `list`(`[10000 + 10000] == [20000]`), `tuple`(`(10000 + 10000,) == (20000,)`), `dict`, `set`, ... . Every complex types shouldn't have to be a value type, but some types are useful to be a value type.

Reply:
Thanks! good information.